core/go/internal/sequencer/coordinator/transaction/pooled.go

```diff
@@ -0,0 +1,139 @@
+		return true
+	}
+
+	if t.PreAssembly.Dependencies != nil {
+		for _, dependencyID := range t.PreAssembly.Dependencies.DependsOn {
+			dependency := t.grapher.TransactionByID(ctx, dependencyID)
+			if dependency == nil {
+				//assume the dependency has been confirmed and no longer in memory
+				//hasUnknownDependencies guard will be used to explicitly ensure the correct thing happens
+				continue
+			}
+			if dependency.isNotAssembled() {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// Function hasUnknownDependencies checks if the transaction has any dependencies the coordinator does not have in memory.  These might be long gone confirmed to base ledger or maybe the delegation request for them hasn't reached us yet. At this point, we don't know
+func (t *Transaction) hasUnknownDependencies(ctx context.Context) bool {
+
+	dependencies := t.dependencies.DependsOn
+	if t.PreAssembly != nil && t.PreAssembly.Dependencies != nil {
+		dependencies = append(dependencies, t.PreAssembly.Dependencies.DependsOn...)
+	}
+
+	for _, dependencyID := range dependencies {
+		dependency := t.grapher.TransactionByID(ctx, dependencyID)
+		if dependency == nil {
+
+			return true
+		}
+
+	}
+
+	//if there are are any dependencies declared, they are all known to the current in memory context ( grapher)
+	return false
+}
+
+// TODO rename this function because it is not clear that its main purpose is to attach this transaction to the dependency as a dependent
+func (t *Transaction) initializeDependencies(ctx context.Context) error {
+	if t.PreAssembly == nil {
+		msg := fmt.Sprintf("cannot calculate dependencies for transaction %s without a PreAssembly", t.ID)
+		log.L(ctx).Error(msg)
+		return i18n.NewError(ctx, msgs.MsgSequencerInternalError, msg)
+	}
+
+	if t.PreAssembly.Dependencies != nil {
+		for _, dependencyID := range t.PreAssembly.Dependencies.DependsOn {
+			dependencyTxn := t.grapher.TransactionByID(ctx, dependencyID)
+
+			if nil == dependencyTxn {
+				//either the dependency has been confirmed and no longer in memory or there was a overtake on the network and we have not received the delegation request for the dependency yet
+				// in either case, the guards will stop this transaction from being assembled but will appear in the heartbeat messages so that the originator can take appropriate action (remove the dependency if it is confirmed, resend the dependency delegation request if it is an inflight transaction)
+
+				//This should be relatively rare so worth logging as an info
+				log.L(ctx).Infof("dependency %s not found in memory for transaction %s", dependencyID, t.ID)
+				continue
+			}
+
+			//TODO this should be idempotent.
+			if dependencyTxn.PreAssembly.Dependencies == nil {
+				dependencyTxn.PreAssembly.Dependencies = &pldapi.TransactionDependencies{}
+			}
+			dependencyTxn.PreAssembly.Dependencies.PrereqOf = append(dependencyTxn.PreAssembly.Dependencies.PrereqOf, t.ID)
```

Comment:
I'm not following how these get updated when the grapher gets reset for complex scenarios. This is a bi-directional dependency on the transactions in a linked list, but there's also the grapher object, and the functions on the grapher object take arrays of transactions for ordering. That's a complex inter-related set of order-management objects on these transactions.

It either needs some very careful explanation and health-warnings (as well as tests) to make sure engineers don't make changes that brake the interrelations....

Or it could just be moved into the grapher and thus simplified.